Fix #2178: testLineAgainstWater works incorrectly outside of game boundaries

Allerek opened the pull request:
Just added a check for water level outside Game Area, and commened out a check if both points are outside Game Area(why would we check for that lol).

Fixes https://github.com/multitheftauto/mtasa-blue/issues/2178

Allerek committed:
no message

Allerek committed:
fixed a typo

forkerer commented:
The commented out check is an early out if both ends are on the same side of the world border, thus skipping all the further calculations, i think that the GetZonesIntersecting return zones at the world border even when both points are ourside of the -3000 : 3000 range
Red arrows show check that will be skipped, while the black one will go through
![obraz](https://user-images.githubusercontent.com/4185198/114630984-082ec680-9cbc-11eb-8e71-62041db7e11c.png)

forkerer commented:
I don't see a reason to remove that check, and i doubt you tested those cases or checked what happens in the code further down

Allerek commented:
> I don't see a reason to remove that check, and i doubt you tested those cases or checked what happens in the code further down

Because thats what this PR is all about, if both of the ends will be outside area, it will automatically return false, we dont want that to happen, we want to be able to check if line is hitting the water in any case, and yeah - i tested that and it worked just fine.

Some users might want to get out of Game Area(for some reason, i'm not here to judge), and we should allow that.

forkerer commented:
And they still can, the code you commented out doesn't change anything about that, it just skips unnecessary calculations later on. No water other than ocean water will ever be collided outside of -3000 : 3000 range because You can't create custom water outside that

forkerer commented:
Check for the GetZonesIntersecting behaviour when you check line starting from (-3050, -3050, 10) to (3050, -3050, 10) when ocean water level is 0, it will do unnecessary checks around the world border if I'm not mistaken after You commented the code out

Allerek commented:
> And they still can, the code you commented out doesn't change anything about that, it just skips unnecessary calculations later on. No water other than ocean water will ever be collided outside of -3000 : 3000 range because You can't create custom water outside that

Yeah, but the code returns false, it doesnt even check if the line really collides with  water, as far as i see it only  checks for X and Y.

-Are both points outside of map?
-If yes then check if they're on the same side of the map
-If so, just return false for some reason? < That doesnt make any sense, whats the reason to suggest theres no water in its way what soever?

forkerer commented:
You literally check for ocean water collision few lines up, no other water will ever exist there, as it can't be created

forkerer commented:
That's the only water you can encounter outside of game zone
![obraz](https://user-images.githubusercontent.com/4185198/114632109-40cf9f80-9cbe-11eb-8b1c-6ac72349e1dd.png)

Allerek commented:
After thinking it out, you might be right, further code checks for waves as i see? Forgot the ocean water doesnt have waves at all, so that can make sense. Getting commented code back, leaving the higher fix tho.

Allerek committed:
Commented code is back.

Allerek commented:
After all the checks are done, i guess we're ready to merge, as the change itself worked well in my tests.

Pirulax commented on Client/game_sa/CWaterManagerSA.cpp:
```diff
@@ -839,12 +839,16 @@ bool CWaterManagerSA::TestLineAgainstWater(const CVector& vecStart, const CVecto
     CVector zeroPoint;
-    if (vecStart.IntersectsSegmentPlane(rayDir, CVector(0, 0, 1), CVector(0, 0, 0), &zeroPoint) && IsPointOutsideOfGameArea(zeroPoint))
+    float   waterHeight;
+    CVector pvecUnknown;
```
Try passing nullptr for `vecUnknown`. It should work (judging from the source code..)

Allerek replied:
I'll try that. Waiting for tests to pass.

Pirulax replied:
I mean the CI will pass, just test it locally, see if everything is as expected.

Allerek committed:
no message

Allerek committed:
Passed null to vector.

Allerek committed:
Changed null to nullptr.

Allerek committed:
Changed zeroPoint name.

Changed zeroPoint name as it is now considering waterHeight, so it's not always returning '0' as Z withing CVector.

Allerek committed:
Changed check for water level to correct.

Allerek commented:
[watertest.zip](https://github.com/multitheftauto/mtasa-blue/files/6351383/watertest.zip)
Test resource.
![image](https://user-images.githubusercontent.com/12304071/115566348-770eaf80-a2ba-11eb-87b9-1309f5a26368.png)
(first number-current water level, second is endz for testLineAgainstWater)

Allerek commented:
For some reason it returns true if water level is lower than 0, and we check at Z=0(or lower), i guess its due to
`GetWaterLevel(CVector(3000,3000,1), &waterHeight, true, nullptr);`
Where i check for Z level '1', i guess we could just set it to a very low number, but this aint the best solution i guess. Any ideas? @forkerer

patrikjuvonen commented:
> For some reason it returns true if water level is lower than 0, and we check at Z=0(or lower), i guess its due to
> `GetWaterLevel(CVector(3000,3000,1), &waterHeight, true, nullptr);`
> Where i check for Z level '1', i guess we could just set it to a very low number, but this aint the best solution i guess. Any ideas? @forkerer

Did you check his suggested fix in the original issue #2178?

Allerek commented:
> > For some reason it returns true if water level is lower than 0, and we check at Z=0(or lower), i guess its due to
> > `GetWaterLevel(CVector(3000,3000,1), &waterHeight, true, nullptr);`
> > Where i check for Z level '1', i guess we could just set it to a very low number, but this aint the best solution i guess. Any ideas? @forkerer
> 
> Did you check his suggested fix in the original issue #2178?

i got some help, i'll introduce it soon but for now i have some work to do

Allerek committed:
Merge branch 'master' of https://github.com/Allerek/mtasa-blue

# Conflicts:
#	Client/game_sa/CWaterManagerSA.cpp

Allerek committed:
Changed CVector bug, thanks to @forkerer.

Allerek committed:
no message

Allerek committed:
no message

Allerek committed:
no message

Allerek committed:
Update CWaterManagerSA.cpp

Allerek committed:
no message

Allerek committed:
no message

Allerek committed:
Merge branch 'master' of https://github.com/Allerek/mtasa-blue

Allerek committed:
no message

Allerek committed:
no message

Pirulax reviewed (changes requested):
LGTM

Pirulax commented on Client/game_sa/CWaterManagerSA.cpp:
```diff
@@ -838,10 +838,14 @@ bool CWaterManagerSA::TestLineAgainstWater(const CVector& vecStart, const CVecto
-        *vecCollision = zeroPoint;
+        *vecCollision = oceanWaterIntersectPoint;
         return true;
     }
```
```suggestion
    // Check for intersection with ocean outside the game area (takes water level into account)
    // If a hit is detected, and it is outside, we early out, as custom water can't be created outside game boundaries
    {
        CVector intersection{};
        const float waterHeight = *static_cast<float*>(0x6E873F);
        if (vecStart.IntersectsSegmentPlane(rayDir, CVector(0, 0, 1), CVector(0, 0, waterHeight), &intersection))
        {
            if (IsPointOutsideOfGameArea(intersection))
            {
                *vecCollision = intersection;
                return true;
            }
        }
    }
```

Allerek committed:
Changes suggested by Pirulax.

Allerek committed:
no message

Allerek committed:
Update CWaterManagerSA.cpp

Pirulax reviewed (approved):
Well I hope you've tested it.
